### PR TITLE
Automatic removal of disabled datasource

### PR DESCRIPTION
**Pull request by moshederri**
Currently, when datasources gets disabled, Turnilo will still query it and keep it visible in the UI.
 As a workaround, one can restart Turnilo service, which eventually ends up refreshing the datasource list.

The changes offered below, remove datasources when they get disabled, by:

1. Removes the datasource when `scanSourceList` gets triggered (only if the datasource is not in the list but exists in the Turnilo's managed datasources).
2. it calls `introspectSources` only on sources that are enabled (same as scanSourceList). This means an additional REST call for sourceList, however, it reduces 3 calls (every `sourceReintrospectInterval` ms) for druid for querying the datasources state (one `segmentMetadata` and 2 `timeBoundary` queries).

### CODE REVIEW

File: src/common/utils/time-monitor/time-monitor.ts [hunk 1]
```diff
@@ -38,6 +38,12 @@ export class TimeMonitor {
     setInterval(this.doChecks, 1000);
   }
 
+  removeCheck(name: string): this {
+    this.checks[name] = null;
```
Comment (adrianmroz) [inline src/common/utils/time-monitor/time-monitor.ts]: Maybe we should make checks more feasible data structure? `Map` or `Set`? With meaningful deletion, so we won't have `!check[name]` condition ...

File: src/server/utils/cluster-manager/cluster-manager.ts [hunk 2]
```diff
@@ -327,6 +335,14 @@ export class ClusterManager {
           if (verbose) logger.log(`For cluster '${cluster.name}' got sources: [${sources.join(", ")}]`);
           // For every un-accounted source: make an external and add it to the managed list.
           let introspectionTasks: Array<Promise<void>> = [];
+
+          this.managedExternals.forEach(ex => {
+            if (sources.filter(src => src === String(ex.external.source)).length === 0) {
```
Comment (adrianmroz) [inline src/server/utils/cluster-manager/cluster-manager.ts]: Use `find`

File: src/common/utils/time-monitor/time-monitor.ts (hunk 1 above)
Comment (moshederri) [reply src/common/utils/time-monitor/time-monitor.ts]: good point, ill change it

File: src/server/utils/cluster-manager/cluster-manager.ts (hunk 2 above)
Comment (moshederri) [reply src/server/utils/cluster-manager/cluster-manager.ts]: ill change it

**Comment by moshederri** [comment]
More about # 2:
Every `sourceReintrospectInterval` ms (configuration in Turnilo config), `introspectSources` is getting executed with the purpose of refreshing the segments metadata. 
The datasources it iterates on do not represent the latest state (for example, when a datasource gets disabled) as it reads the list from an in-memory variable (managedExternals) which gets updated (by 'scanSourceList') on a different user-configurable interval.
One possible solution is to make both functions (`scanSourceList`/`introspectSources`) modify the `managedExternals` list, however, it requires locking.
The other solution I thought of (having fewer changes possible) is rather than rely on the correctness of `managedExternals`, is to query druid for the current datasource list (simple REST request which happens not that often) and on it,  check for changes in segments. 
 Not doing this change will end up issuing 2 invalid queries (`segmentMetadata` + `timeBoundary`) every time it tries to refresh the segments metadata.
In addition, I found a bug where Turnilo issues more than 10K requests per minute (messing our performance tests) every time a user disableds a datasources. I'll issue the fix soon).

File: src/server/utils/cluster-manager/cluster-manager.ts [hunk 3]
```diff
@@ -370,9 +386,25 @@ export class ClusterManager {
...
 
-    return Promise.all(this.managedExternals.map(managedExternal => {
-      return this.introspectManagedExternal(managedExternal);
-    })).then(noop);
+    return (External.getConstructorFor(cluster.type) as any).getSourceList(this.requester)
+        .then(
+            (sources: string[]) => {
+              let introspectionTasks: Array<Promise<void>> = [];
+              sources.forEach(source => {
+                const existingExternalsForSource = this.managedExternals.filter(managedExternal => getSourceFromExternal(managedExternal.external) === source);
```
Comment (adrianmroz) [inline src/server/utils/cluster-manager/cluster-manager.ts]: When source could have more than one external?

**Comment by adrianmroz** [comment]
> In addition, I found a bug where Turnilo issues more than 10K requests per minute (messing our performance tests) every time a user disableds a datasources. I'll issue the fix soon).

Any updates on that? Is it existing bug in turnilo or something that came out with your PR?

File: src/server/utils/cluster-manager/cluster-manager.ts (hunk 3 above)
Comment (moshederri) [reply src/server/utils/cluster-manager/cluster-manager.ts]: I wondered about it as well but I couldn't find the answer.
The method which it calls is `getSourceList` which comes from plywood lib 
(https://github.com/implydata/plywood/blob/master/src/external/druidExternal.ts)
Keep in mind that this method was used in the code earlier and not something new.

**Comment by moshederri** [comment]
Well, the bug (which is easily reproduced - just disable a datasource and see a spike in the segmentMetadata/timeBoundary queries) does not remove the disabled datasources from the code which checks whether a change was made to the datasource.

The code that performs the check can be found here:
https://github.com/allegro/turnilo/blob/master/src/common/utils/time-monitor/time-monitor.ts

In a nutshell, it checks whether to refresh the datasource metadata every 1000 by checking the last time the datasource got refreshed (section in the code called `timeTag`).
However, when the datasource gets disabled, the refresh code fails and eventually does not update the timeTag (leading to a constant refresh every 1000 ms). 

Bottom line:
This PR removes this datasouce from the checklist as well (which means it doesn't get refreshed). However, Turnilo will still issue refresh requests after a datasource gets disabled until `sourceListRefreshInterval` gets fired (not a critical issue - but could be fixed as well).
If this PR will not get be merged, then I would create an additional PR fixing it.

**Comment by moshederri** [comment]
@adrianmroz, is there an update when/if this fix will be merged?

**Comment by adrianmroz** [comment]
Soon :) I was on winter break and I'm getting back on track.

**Comment by alexbusu** [comment]
@moshederri , does this work for `string[]` source as well? I have an union in my config, where two actual sources are declared; that cube doesn't show up on Turnilo dashboard.
My cluster config contains this
```
sourceListScan: disable
sourceListRefreshOnLoad: false
sourceReintrospectOnLoad: false
```
and 
```
dataCubes:
- name: union
  title: union
  clusterName: druid
  source: ["redirects", "sales"]
  refreshRule:
    rule: 'realtime'
  introspection: autofill-all
```

Is there something wrong in my configuration, or there's room for improvement for your bugfix?
Thanks!
cc @adrianmroz

**Comment by adrianmroz** [comment]
I see there's comparison with `String` constructor, so probably your array is converted do something like `["redirects", "sales"]` and doesn't match. 

Question is - how this should behave? If "sales" goes out, should cube "union" be disabled altogether? Probably yes.

If I understand your report - you have this config and after this change (or some change in between) you config stopped working i.e. you don't see union cube in your list? If you run turnilo with `--verbose` flag, do you get any warnings?

**Comment by alexbusu** [comment]
> I see there's comparison with `String` constructor, so probably your array is converted do something like `["redirects", "sales"]` and doesn't match.

In this case maybe the string should be parsed as... JSON? And then proceed accordingly depending on the output type.

> Question is - how this should behave? If "sales" goes out, should cube "union" be disabled altogether? Probably yes.

Yes, I agree, if one source is unavailable, then the cube gets deactivated.


These are the last lines I get when starting Turnilo
```
turnilo    | vvvvvvvvvvvvvvvvvvvvvvvvvv
turnilo    | Requester rq69364 sending query 2:
turnilo    | {
turnilo    |   "queryType": "sourceList"
turnilo    | }
turnilo    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
turnilo    | vvvvvvvvvvvvvvvvvvvvvvvvvv
turnilo    | Requester rq69364 got result from query 2: (in 110ms)
turnilo    | [
turnilo    |   [
turnilo    |     "sales",
turnilo    |     "redirects",
turnilo    |     "external-sales",
turnilo    |     "external-redirects",
turnilo    |     "events"
turnilo    |   ]
turnilo    | ]
turnilo    | ^^^^^^^^^^^^^^^^^^^^^^^^^^

```

**Comment by adrianmroz** [comment]
> In this case maybe the string should be parsed as... JSON? And then proceed accordingly depending on the output type.

We have source as a value at hand, so we should just use equality for strings and `Array.find` for arrays.

> These are the last lines I get when starting Turnilo
> 
> ```
> turnilo    | vvvvvvvvvvvvvvvvvvvvvvvvvv
> turnilo    | Requester rq69364 sending query 2:
> turnilo    | {
> turnilo    |   "queryType": "sourceList"
> turnilo    | }
> turnilo    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

And that's concerning! why `sourceList` was sent when you have disabled it? I assume after receiving update cluster manager removes your cube because of this comparison issue above.

**Comment by alexbusu** [comment]
> I assume after receiving update cluster manager removes your cube because of this comparison issue above.

I guess so. This is the missing part from the Turnilo startup (the first log lines)
```
turnilo    | Using config config.yaml
turnilo    | Starting Turnilo v1.21.0
turnilo    | Turnilo is listening on address :: port 9090
turnilo    | Adding cluster manager for 'druid' with 1 dataCube
turnilo    | Attempting to connect to cluster 'druid'
turnilo    | vvvvvvvvvvvvvvvvvvvvvvvvvv
turnilo    | Requester rq69364 sending query 1:
turnilo    | {
turnilo    |   "queryType": "status"
turnilo    | }
turnilo    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
turnilo    | vvvvvvvvvvvvvvvvvvvvvvvvvv
turnilo    | Requester rq69364 got result from query 1: (in 980ms)
turnilo    | [
turnilo    |   {
turnilo    |     "version": "0.13.0-incubating",
turnilo    |     "modules": [ ... ],
turnilo    |     "memory": {
turnilo    |       "maxMemory": 31999393792,
turnilo    |       "totalMemory": 31999393792,
turnilo    |       "freeMemory": 30731741840,
turnilo    |       "usedMemory": 1267651952
turnilo    |     }
turnilo    |   }
turnilo    | ]
turnilo    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
turnilo    | Connected to cluster 'druid'
turnilo    | Cluster 'druid' is running druid@0.13.0-incubating
turnilo    | Introspecting all sources in cluster 'druid'
[then is the part I posted above]
```

**Comment by alexbusu** [comment]
It is a comma-separated list.
image

**Comment by adrianmroz** [comment]
That function should be fixed, so it can return `string | string[]`. 

Did you debug why it starts at all? Shouldn't this condition short circuit: https://github.com/allegro/turnilo/blob/master/src/server/utils/cluster-manager/cluster-manager.ts#L329 ?

This line then fires and condition is not met: https://github.com/allegro/turnilo/blob/master/src/server/utils/cluster-manager/cluster-manager.ts#L340 `String(ex.external.source)` probably evaluates to something like above and don't match.

**Comment by alexbusu** [comment]
I think I managed to have a working solution. Should I PR on master (target of 1.24)?

**Comment by alexbusu** [comment]
> Shouldn't this condition short circuit: https://github.com/allegro/turnilo/blob/master/src/server/utils/cluster-manager/cluster-manager.ts#L329 ?

After I applied the patch, that scan doesn't fire anymore.

**Comment by alexbusu** [comment]
Ok, that api call actually happens all the time. It is originated in `ClusterManager.introspectSources()` where plywood's `DruidExternal.getSourceList()` is called.
